repose-aggregator/tests/functional-tests/src/integrationTest/groovy/features/services/datastore/DistDatastoreClientAuthMultiTest.groovy

```diff
@@ -0,0 +1,138 @@
+
+import framework.ReposeValveTest
+import org.apache.http.HttpResponse
+import org.apache.http.conn.ssl.NoopHostnameVerifier
+import org.apache.http.conn.ssl.SSLConnectionSocketFactory
+import org.apache.http.impl.client.CloseableHttpClient
+import org.apache.http.impl.client.HttpClients
+import org.apache.http.ssl.SSLContexts
+import org.apache.http.util.EntityUtils
+import org.openrepose.commons.utils.io.ObjectSerializer
+import org.openrepose.core.services.datastore.types.StringValue
+import org.rackspace.deproxy.*
+import spock.lang.Shared
+
+import javax.net.ssl.SSLContext
+import java.nio.file.Files
+import java.util.concurrent.TimeUnit
+
+class DistDatastoreClientAuthMultiTest extends ReposeValveTest {
+    //Since we're serializing objects here for the dist datastore, we must have the dist datastore objects in our classpath
+    final ObjectSerializer objectSerializer = new ObjectSerializer(this.getClass().getClassLoader())
+
+    @Shared
+    def params
+    @Shared
+    def String distDatastoreEndpoint
+    @Shared
+    def File serverFile
+    @Shared
+    def File clientFile
+
+    def setupSpec() {
+        int dataStorePort = PortFinder.Singleton.getNextOpenPort()
+
+        distDatastoreEndpoint = "https://localhost:${dataStorePort}"
+
+        params = properties.getDefaultTemplateParams()
+        params += [
+                'datastorePort': dataStorePort
+        ]
+
+        reposeLogSearch.cleanLog()
+        repose.configurationProvider.cleanConfigDirectory()
+
+        repose.configurationProvider.applyConfigs("common", params)
+        // Have to manually copy binary files, because the applyConfigs() attempts to substitute template parameters
+        // when they are found and it breaks everything. :(
+        def serverFileOrig = new File(repose.configurationProvider.configTemplatesDir, "common/server.jks")
+        serverFile = new File(repose.configDir, "server.jks")
+        def serverFileDest = new FileOutputStream(serverFile)
+        Files.copy(serverFileOrig.toPath(), serverFileDest)
+        def clientFileOrig = new File(repose.configurationProvider.configTemplatesDir, "common/client.jks")
+        clientFile = new File(repose.configDir, "client.jks")
+        def clientFileDest = new FileOutputStream(clientFile)
+        Files.copy(clientFileOrig.toPath(), clientFileDest)
+        repose.configurationProvider.applyConfigs("features/services/datastore/clientauth", params)
+        repose.configurationProvider.applyConfigs("features/services/datastore/clientauth/truststore", params)
+
+        SSLContext sslContext = SSLContexts.custom()
+                .loadKeyMaterial(serverFile, "password".toCharArray(), "password".toCharArray())
+                .loadTrustMaterial(clientFile, "password".toCharArray())
+                .build()
+        CloseableHttpClient client = HttpClients.custom()
+                .setSSLSocketFactory(new SSLConnectionSocketFactory(sslContext, NoopHostnameVerifier.INSTANCE))
```

Comment:
We want to add a setting to the connection pool for hostname verification (allowing it to be disabled, at least). This will, of course, require additional documentation.